text/0000-duration-reform.md

```diff
@@ -0,0 +1,162 @@
+
+Human complications such as leap seconds, days in a month, and leap years, and machine complications such as NTP adjustments make these concepts and their full APIs more complicated than they would at first appear. This proposal focuses on fleshing out a design for `Duration` that is sufficient for use as a timeout, leaving the other concepts of time to a future proposal.
+
+---
+
+For the most part, the system APIs that this type is used to communicate with either use `timespec` (`u64` seconds plus `u32` nanos) or take a timeout in milliseconds (`u32` on Windows).
+
+> For example, [`GetQueuedCompletionStatus`][iocp-ms-example], one of
+> the primary APIs in the Windows IOCP API, takes a `dwMilliseconds`
+> parameter as a [`DWORD`][msdn-dword], which is a `u32`. Some Windows
+> APIs use "ticks" or 100-nanosecond units.
+
+[iocp-ms-example]: https://msdn.microsoft.com/en-us/library/windows/desktop/aa364986%28v=vs.85%29.aspx
+[msdn-dword]: https://msdn.microsoft.com/en-us/library/cc230318.aspx
+
+In light of that, this proposal has two primary goals:
+
+* to define a type that can describe portable timeouts for cross-
+  platform APIs
+* to describe what should happen if a large `Duration` is passed into
+  an API that does not accept timeouts that large
+
+In general, this proposal considers it acceptable to reduce the granularity of timeouts (eliminating nanosecond granularity if only milliseconds are supported) and to truncate very large timeouts.
+
+This proposal retains the two fields in the existing `Duration`:
+
+* a `u64` of seconds
+* a `u32` of additional nanosecond precision
+
+Timeout APIs defined in terms of milliseconds will truncate `Duration`s that are more than `u32::MAX` in milliseconds, and will reduce the granularity of the nanosecond field.
+
+> A `u32` of milliseconds supports a timeout longer than 45 days.
+
+Future APIs to support a broader set of [Durations][joda-duration] APIs, a [Period][joda-period] and [Instant][joda-instant] type, as well as coercions between these types, would be useful, compatible follow-ups to this RFC.
+
+[joda-duration]: http://www.joda.org/joda-time/key_duration.html
+[joda-period]: http://www.joda.org/joda-time/key_period.html
+[joda-instant]: http://www.joda.org/joda-time/key_instant.html
+
+# Detailed design
+
+A `Duration` represents a period of time represented in terms of nanosecond granularity. It has `u64` seconds and an additional `u32` nanoseconds. There is no concept of a negative `Duration`.
+
+> A negative `Duration` has no meaning for many APIs that may wish
+> to take a `Duration`, which means that all such APIs would need
+> to decide what to do when confronted with a negative `Duration`.
+> As a result, this proposal focuses on the predominant use-cases for
+> `Duration`, where unsigned types remove a number of caveats and
+> ambiguities.
+
+```rust
+pub struct Duration {
+  secs: u64,
+  nanos: u32 // may not be more than 1 billion
```

Comment:
Does this mean it is allowed to be exactly 1 billion, or is only 999,999,999 allowed?